docs/specs/deployment-wise-sms-gateway.md

```diff
@@ -0,0 +1,89 @@
+
+## Behavior explained
+
+### Table 1
+
+| `sms_provider` | `sms_gateway` | Behavior                                                                                                                                |
+| -------------- | ------------- | --------------------------------------------------------------------------------------------------------------------------------------- |
+| absent         | absent        | See table 2.                                                                                                                            |
+| absent         | **present**   | Use sms gateway config. See Table 3                                                                                                     |
+| **present**    | absent        | Use `messaging.sms_provider` from `authgear.secrets.yaml`. Read config from `sms.{messaging.sms_provider}` from `authgear.secrets.yaml` |
+| **present**    | **present**   | `messaging.sms_provider` will be ignored. Use sms gateway config, See Table 3                                                           |
+
+### Table 2
+
+When `sms_provider` and `sms_gateway` are absent, the provider config is termined by environment variables.
+
+| `sms_provider` | `DEFAULT_USE_CONFIG_FROM` | `DEFAULT_PROVIDER` | Behavior                                                                                                                                   |
+| -------------- | ------------------------- | ------------------ | ------------------------------------------------------------------------------------------------------------------------------------------ |
+| absent         | absent                    | NC                 | `provider` will be determined from application logic. Read config from `sms.{provider}` from `authgear.secrets.yaml`                       |
+| absent         | `environment_variable`    | absent             | `provider` will be determined from application logic. Read config from `SMS_GATEWAY_{provider}_*` from environment variables               |
+| absent         | `environment_variable`    | **present**        | Use `SMS_GATEWAY_DEFAULT_PROVIDER` as provider. Will read config from `SMS_GATEWAY_{SMS_GATEWAY_DEFAULT_PROVIDER}_*` environment variables |
+| absent         | `authgear.secrets.yaml`   | NC                 | `provider` will be determined from application logic. Read config from `sms.{provider}` from `authgear.secrets.yaml`                       |
+| **present**    | NC                        | NC                 | `messaging.sms_provider` presents. See Table 1                                                                                             |
+
+### Table 3
+
+When `sms_gateway` is provided,
+
+| `sms_gateway` | `sms_gateway.use_config_from` | `sms_gateway.provider` | `DEFAULT_PROVIDER` | Behaviour                                                                                                                                  |
```

Comment:
Add the column `DEFAULT_USE_CONFIG_FROM`. You did not explain the behavior of `sms_gateway` is present, and the interaction with `DEFAULT_USE_CONFIG_FROM`

Reply:
Since `sms_gateway.use_config_from` is required when `sms_gateway` is presented in authgear.yaml, it should always override `DEFAULT_USE_CONFIG_FROM` in environment variable.

I have added this note to the doc.